src/moi.jl

```diff
@@ -289,6 +289,88 @@ end
 accept_vector_set(::ProductMode{T}, ::Complement) where {T} = nothing
 accept_vector_set(::MixedMode{T}, ::Complement) where {T} = nothing
 
+function get_variable_complement(primal_model, dual_model, primal_con, dual_con)
+    return error(
+        "An internal error with variable complements occurred. Likely, your problem type does not yet support consideration of constrained variables.",
+    )
+end
+
+function get_variable_complement(
+    primal_model,
+    dual_model,
+    primal_con::MOI.ConstraintIndex{Fp,Sp},
+    dual_con::MOI.ConstraintIndex{Fd,Sd},
+) where {
+    Fp<:MOI.VariableIndex,
+    Sp<:Union{MOI.LessThan{T},MOI.GreaterThan{T}},
+    Fd,
+    Sd<:Union{MOI.LessThan{T},MOI.GreaterThan{T}},
+} where {T}
+    primal_variable =
+        MOI.get(primal_model, MOI.ConstraintFunction(), primal_con)
+    primal_set = MOI.get(primal_model, MOI.ConstraintSet(), primal_con)
+
+    @assert MOI.constant(primal_set) == 0 "Unexpected variable bound"
+
+    dual_func = MOI.get(dual_model, MOI.ConstraintFunction(), dual_con)
+    dual_set = MOI.get(dual_model, MOI.ConstraintSet(), dual_con)
+    if MOI.constant(dual_set) > 0
+        dual_func.constant = dual_func.constant - MOI.constant(dual_set)
+    elseif MOI.constant(dual_set) < 0
+        dual_func.constant = dual_func.constant + MOI.constant(dual_set)
+    end
+    return Complement(
+        false,
+        primal_con,
+        dual_func,
+        set_with_zero(dual_set),
+        primal_variable,
+    )
+end
+
+function get_variable_complement(
+    primal_model,
+    dual_model,
+    primal_con::MOI.ConstraintIndex{Fp,Sp},
+    dual_con::MOI.ConstraintIndex{Fd,Sd},
+) where {Fp<:MOI.VectorOfVariables,Sp<:VECTOR_SETS,Fd,Sd<:VECTOR_SETS} where {T}
+    primal_variables = MOI.copy(
+        MOI.get(primal_model, MOI.ConstraintFunction(), primal_con),
+    )::Fp
+    primal_set =
+        MOI.copy(MOI.get(primal_model, MOI.ConstraintSet(), primal_con))::Sp
+
+    dual_func =
+        MOI.copy(MOI.get(dual_model, MOI.ConstraintFunction(), dual_con))
+    dual_set = MOI.copy(MOI.get(dual_model, MOI.ConstraintSet(), dual_con))
+
+    # Do we have a todo here as in function get_canonical_complement(primal_model, map, ci::CI{F,S}) where {F, S<:VECTOR_SETS} ??
```

Comment:
do what?

Reply:
There was the following todo in the code of get_canonical_complement (note that I commented one additional line out):
```julia
    # dim = MOI.dimension(set)
    # vector sets have no constant
    # for i in 1:dim
    #     func.constant[i] = Dualization.set_dot(i, set, T) *
    #         Dualization.get_scalar_term(primal_model, ci, i)
    # end
    # todo - set dot on function
```